application/core/plugins/GoogleOAuthSMTP/GoogleOAuthSMTP.php

```diff
@@ -0,0 +1,484 @@
+            $this->set('refreshToken', null);
+            $this->set('refreshTokenMetadata', []);
+            $this->set('email', null);
+
+            // If credentials are complete, we redirect to settings page so the user can
+            // retrieve the token from Google.
+            if (!empty($clientId) && !empty($clientSecret)) {
+                Yii::app()->user->setFlash('success', gT('The plugin settings were saved.'));
+                $this->redirectToConfig();
+            }
+        }
+    }
+
+    /**
+     * Override mailer
+     */
+    public function beforeEmail()
+    {
+        if (!$this->get('enable')) {
+            return;
+        }
+
+        try {
+            [$clientId, $clientSecret] = $this->getCredentials();
+        } catch (Exception $ex) {
+            $this->log("GoogleOAuthSMTP is enabled but credentials are incomplete.", CLogger::LEVEL_WARNING);
+            return;
+        }
+
+        $refreshToken = $this->get("refreshToken");
+        if (empty($refreshToken)) {
+            $this->log("GoogleOAuthSMTP is enabled but there is no refresh token stored.", CLogger::LEVEL_WARNING);
+            return;
+        }
+
+        $emailAddress = $this->get("email");
+        if (empty($emailAddress)) {
+            $this->log("GoogleOAuthSMTP is enabled but there is no email address. Please generate a new token.", CLogger::LEVEL_WARNING);
+            return;
+        }
+
+        $limeMailer = $this->getEvent()->get('mailer');
+        $limeMailer->IsSMTP();
+        $limeMailer->Host = 'smtp.gmail.com';
+        $limeMailer->Port = 465;
+        $limeMailer->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS;
+        $limeMailer->SMTPAuth = true;
+        $limeMailer->Username = null;
+        $limeMailer->Password = null;
+        $limeMailer->AuthType = 'XOAUTH2';
+
+        $provider = $this->getProvider($clientId, $clientSecret);
+        $limeMailer->setOAuth(new OAuth(
+            [
+                'provider' => $provider,
+                'clientId' => $clientId,
+                'clientSecret' => $clientSecret,
+                'refreshToken' => $refreshToken,
+                'userName' => $emailAddress,
+            ]
+        ));
+
+        // Set "Reply To" because Gmail overrides the From/Sender with the logged user.
+        $limeMailer->AddReplyTo($limeMailer->From, $limeMailer->FromName);
+    }
+
+    /**
+     * @inheritdoc
+     * Block irrelevant settings on Global Settings screen.
+     */
+    public function beforeControllerAction()
+    {
+        if (!$this->get('enable')) {
+            return;
+        }
+
+        $controller = $this->getEvent()->get('controller');
+        $action = $this->getEvent()->get('action');
+
+        if ($controller == 'admin' && $action == 'globalsettings') {
+            $this->registerGlobalSettingsScript();
+        }
+    }
+
+    private function registerGlobalSettingsScript()
+    {
```

Comment:
@olleharstedt about this hack to update Global Settings form, you don't work on something allowing update any form currently ?

Reply:
Yes, it's already available. https://manual.limesurvey.org/Plugins_-_advanced#Form_extension_.28New_in_6_.29

Might need to be extended, of course.

Reply:
But you can't use it to disable items, as they do here.

Reply:
The idea is not to use a hack nor a plugin form event, but to enhance the core as to allow Oauth Email plugins.